content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,232 @@ Like a segment tree on values rather than indices.
+- $prefixB[i]$ as prefix sum of $c[i]$
+
+Formally
+
+$$
+c[i] = index[i] < mid ? 1 : 0;
+prefixB[i] = prefixB[i - 1] + c[i]
+$$
+
+
+To update $r$ as we recurse down, we do the following:
+- To know the value of 𝑟 if we recurse left, we use prefixB[r]
+- If we recurse right, we use 𝑟 - prefixB[r]
+
+## Solving the second type of query
+**Given a range l, r find the k smallest element**
+
+We will determine whether the answer for a given node is in the left or the
+right segment.
+We can calculate how many times the elements within the segments' ranges appear
+in our range $(l, r)$ using our first type of query.
+Note that this also works for non-leaf nodes using the following formula:
+
+$$
+\texttt{occurrences}(l, r) = r - l
+$$
+<Info title="Simular">
+This is similar to counting how many times a value appears up to index 𝑅 in our previous query. We did this by using the new 𝑅 value at the leaf node. But now, we consider the difference between the updated 𝑅 and 𝐿
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+\texttt{left\_occurrences} = prefixB[r] - prefixB[l]
+$$
+
+<Info title="Left occurrences">
+  Note that $\texttt{left\_occurrences}$ is the number of indices between l and r whose value is less than mid
+
+</Info>
+
+- If $\texttt{left\_occurrences}$ is greater or equal to $k$, it means the $k$-th smallest element is in
+  the left subtree. Therefore, we update our range and recurse into the left
+  child
+- If $\texttt{left\_occurrences}$ is less than $k$, it means the
+  $k$-th smallest element is in the right subtree. We adjust k by subtracting
+  $\texttt{left\_occurrences}$ from $k$, update our range, and recurse into the right child
+
+<Info title="Notice">
+  Notice we still update $l, r$ accordingly when we go left or right
+</Info>
+
+the answer then will be the value of the node we end up on (leaf)
+
+## Implemention
+**Time Complexity:** $\mathcal{O}(Q \cdot \log(M))$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+struct Segment {
+	Segment *left = nullptr, *right = nullptr;
+	int l, r, mid;
+	bool children = false;
+	vector<pair<int, int>> indices;  // Index, Value
+	vector<int> prefixB;
+
+	Segment(int l, int r, const vector<pair<int, int>> &indices)
+	    : l(l), r(r), mid((r + l) / 2), indices(indices) {
+		calculate_b();
+	}
+
+	// Sparse since values can go up to 1e9
+	void update() {
+		if (children) { return; }
+		children = true;
+		if (r - l > 1) {
+			// Split the indices for left and right child
+			vector<pair<int, int>> leftIndices, rightIndices;
+			partition_copy(indices.begin(), indices.end(), leftIndices.begin(),
+			               rightIndices.begin(), [this](const pair<int, int> &elem) {
+				               return elem.second < mid;
+			               });
+
+			left = new Segment(l, mid, leftIndices);
+			right = new Segment(mid, r, rightIndices);
+		}
+	}
+
+	// Calculates the prefix B
+	void calculate_b() {
```

Comment:
why not put this in the constructor?

Reply:
The idea was to split the code into pieces to make it easier to understand, would you prefer it inside the constructor?

Reply:
you could like put a comment inside the constructor saying ur calculating the prefix of b, then put the calculate_b code inside the constructor

Reply:
By splitting the code into functions, we reduce future code duplication and enhance maintainability, even if reuse isn't immediately necessary.

Reply:
idk man i think this code would be just as readable/maintainable if you put this inside the constructor with a comment saying what it does